File: go/graphs/adj-mat.go
package main

import "fmt"
import "container/list"

func printGraph(nodes [5][5]int) {
    for i := 0; i < len(nodes); i++ {
        fmt.Printf("(%d)", i)
        for j := 0; j < len(nodes); j++ {
            if nodes[i][j] > 0 {
                fmt.Printf(" -%d> %d", nodes[i][j], j)
            }
        }
        fmt.Println()
    }
}

func BellmanFord(nodes [5][5]int, pn int) ([]int) {
    fila := list.New()
    dist := make([]int, len(nodes))
    for i := 0; i < len(dist); i++ {
        dist[i] = 100;
    }
    procesado := make([]bool, len(nodes))

    fila.PushBack(pn)
    dist[pn] = 0

    for fila.Len() > 0 {
        act := fila.Remove(fila.Front())
        for j := 0; j < len(nodes); j++ {
            if nodes[act.(int)][j] > 0 {
                if dist[act.(int)] + nodes[act.(int)][j] <  dist[j] {
                    dist[j] = dist[act.(int)] + nodes[act.(int)][j]
                }
                if !procesado[j] {
                    fila.PushBack(j)
                }
            }
        }
        procesado[act.(int)] = true
    }

    return dist
}

func main() {
    var nodes [5][5]int
    nodes[0][1] = 1
    nodes[0][2] = 15
    nodes[1][3] = 6
    nodes[1][4] = 2
    nodes[3][1] = 1
    nodes[3][2] = 5
    nodes[4][3] = 3

    printGraph(nodes)
    distances := BellmanFord(nodes, 0)
    for d := 0; d < len(distances); d++ {
        fmt.Printf("De 0 a %d: %d\n", d, distances[d])
    }
}

File: rust/graphs/adj-mat.rs
use std::collections::VecDeque;

struct Graph {
    nodes: Vec<Vec<usize>>,
}

impl Graph {
    fn new(pn: usize) -> Self {
        let mut tv: Vec<Vec<usize>> = Vec::new();
        for _i in 0..pn {
            let mut v: Vec<usize> = Vec::new();
            for _j in 0..pn {
                v.push(0);
            }
            tv.push(v);
        }
        Self {
            nodes: tv,
        }
    }
    fn add_edge(&mut self, ini: usize, fin: usize, peso: usize) {
        self.nodes[ini][fin] = peso;
    }
    fn print_graph(&self) {
        for i in 0..self.nodes.len() {
            print!("({})",i);
            for j in 0..self.nodes.len() {
                if self.nodes[i][j] > 0 {
                    print!(" -{}> {}", self.nodes[i][j], j);
                }
            }
            println!("");
        }
    }
    fn bellman_ford(&self, pn: usize, dist: &mut [usize]) {
        let mut fila = VecDeque::new();
        let mut procesado = vec![false;self.nodes.len()];

        fila.push_front(pn);
        dist[pn] = 0;

        while fila.len() > 0 {
            match fila.pop_front() {
                Some(act) => {
                    for j in 0..self.nodes.len() {
                        if self.nodes[act][j] > 0 {
                            if dist[act] + self.nodes[act][j] < dist[j] {
                                dist[j] = dist[act] + self.nodes[act][j];
                            }
                            if !procesado[j] {
                                fila.push_back(j);
                            }
                        }
                    }
                    procesado[act] = true;
                },
                None => (),
            }
        }
    }
}

fn main() {
    let mut g = Graph::new(5);
    g.add_edge(0, 1, 1);
    g.add_edge(0, 2, 15);
    g.add_edge(1, 3, 6);
    g.add_edge(1, 4, 2);
    g.add_edge(3, 1, 1);
    g.add_edge(3, 2, 5);
    g.add_edge(4, 3, 3);

    g.print_graph();
    let mut dist:[usize;5] = [100;5];
    g.bellman_ford(0, &mut dist);
    for d in 0..dist.len() {
        println!("De 0 a {}: {}", d, dist[d]);
    }
}
